stdlib/public/Concurrency/Clock.cpp

```diff
@@ -52,14 +56,16 @@ void swift_get_time(
       QueryPerformanceFrequency(&freq);
       LARGE_INTEGER count;
       QueryPerformanceCounter(&count);
-      *seconds = count.QuadPart / freq.QuadPart;
-      if (freq.QuadPart < 1000000000) {
-        *nanoseconds = 
-            ((count.QuadPart % freq.QuadPart) * 1000000000) / freq.QuadPart;
-      } else {
-        *nanoseconds = 
-            (count.QuadPart % freq.QuadPart) * (1000000000.0 / freq.QuadPart);
-      }
+      // Divide count (number of ticks) by frequency (number of ticks per
+      // second) to get the counter in seconds. We also need to multiply the
+      // count by 1,000,000,000 to get nanosecond resolution. By multiplying
+      // first, we maintain high precision. The resulting value is the tick
+      // count in nanoseconds. Use 128-bit math to avoid overflowing.
+      DWORD64 hi = 0;
+      DWORD64 lo = _umul128(count.QuadPart, 1'000'000'000, &hi);
+      DWORD64 ns = _udiv128(hi, lo, freq.QuadPart, nullptr);
```

Comment:
`_udiv128()` generates a GPF if the result can't fit in a 64-bit integer, but if that happens that means the computer's been up for more than 2^64 nanoseconds… that's 584 years. If anybody runs into that problem, please let my great-great-great-great-great-great-great-great-great-great-great-great-great-great-great-great-great-great-great-great-great-great-great-great-great-great-grandkid know.